apps/indexer/src/api/services/historical-balances/historical-balances.service.ts

```diff
@@ -0,0 +1,272 @@
+  private async getBalancesIndividually(
+    addresses: Address[],
+    blockNumber: number,
+    tokenAddress: Address
+  ): Promise<HistoricalBalance[]> {
+    const balancePromises = addresses.map((address) =>
+      readContract(this.client, {
+        address: tokenAddress,
+        abi: [
+          {
+            inputs: [{ name: "account", type: "address" }],
+            name: "balanceOf",
+            outputs: [{ name: "", type: "uint256" }],
+            stateMutability: "view",
+            type: "function",
+          },
+        ] as const,
+        functionName: "balanceOf",
+        args: [address],
+        blockNumber: BigInt(blockNumber),
+      }).catch((error) => {
+        console.error(`Error fetching balance for ${address}:`, error);
+        return 0n; // Return 0 if individual call fails
+      })
+    );
+
+    const balances = await Promise.all(balancePromises);
+
+    // Transform results into HistoricalBalance objects
+    return addresses.map((address, index) => ({
+      address: address!,
+      balance: balances[index] || 0n,
+      blockNumber,
+      tokenAddress,
+    }));
+  }
+
+  /**
+   * Maps DAO ID to corresponding token contract address
+   */
+  private getTokenAddress(daoId: DaoIdEnum): Address | null {
+    const { NETWORK: network } = env;
+    const contractInfo = CONTRACT_ADDRESSES[network]?.[daoId];
+    return contractInfo?.token?.address || null;
+  }
+
+  /**
+   * Validate the historical balances request
+   * @param request - The request to validate
+   */
+  private validateRequest(request: HistoricalBalancesRequest): void {
+    const { addresses, blockNumber, daoId } = request;
+
+    // Validate addresses array
+    if (!Array.isArray(addresses) || addresses.length === 0) {
+      throw new Error("Addresses array cannot be empty");
+    }
+
+    if (addresses.length > 100) {
+      throw new Error("Maximum 100 addresses allowed per request");
+    }
+
+    // Validate block number
+    if (!Number.isInteger(blockNumber) || blockNumber <= 0) {
+      throw new Error("Block number must be a positive integer");
+    }
+
+    // Validate DAO ID
+    if (!Object.values(DaoIdEnum).includes(daoId)) {
+      throw new Error(`Invalid DAO ID: ${daoId}`);
+    }
+
+    // Validate that addresses are valid Ethereum addresses
+    addresses.forEach((address, index) => {
+      if (!this.isValidAddress(address)) {
+        throw new Error(`Invalid address at index ${index}: ${address}`);
+      }
+    });
+  }
+
+  /**
+   * Get token contract information for a specific DAO
+   * @param daoId - The DAO identifier
+   * @returns Token contract address and ABI
+   */
+  public getTokenContract(daoId: DaoIdEnum) {
+    const { NETWORK: network } = env;
+    const contractInfo = CONTRACT_ADDRESSES[network][daoId];
+
+    if (!contractInfo?.token) {
+      throw new Error(
+        `Token contract not found for DAO: ${daoId} on network: ${network}`,
+      );
+    }
+
+    return contractInfo.token;
+  }
```

Comment:
this should be injected as argument of this service method instead of relying on the env within the inner layers of the application